# 주어진 DataSet으로 이진 분류 구현
소득 $50000 초과 여부 분류   

Binary Classification for over or below $50000 income

In [1]:
import os
import pandas as pd
import numpy as np

## I. 작업 경로 설정

In [2]:
# 변경 전 확인
print(os.getcwd())

# 작업 경로 변경
os.chdir('/Users/ryu/Desktop/데스크탑 - ryuseungho의 MacBook Air/2022/Bigdata/Machine-Learning/Project_data')

#Binary_Classification

# 변경 후 확인
print(os.getcwd())

/Users/ryu/Machine_Learning_Project1
/Users/ryu/Desktop/데스크탑 - ryuseungho의 MacBook Air/2022/Bigdata/Machine-Learning/Project_data


## II. 데이터 불러오기

In [3]:
adult_data = pd.read_csv('./Binary_Classification/adult_data.csv', sep=',')
adult_test = pd.read_csv('./Binary_Classification/adult_test.csv', sep=',')

## III. 데이터 전처리 

### adult_data 전처리

In [4]:
adult_data

age          workclass  fnlwgt    education  eduaction-num  \
0       39          State-gov   77516    Bachelors             13   
1       50   Self-emp-not-inc   83311    Bachelors             13   
2       38            Private  215646      HS-grad              9   
3       53            Private  234721         11th              7   
4       28            Private  338409    Bachelors             13   
...    ...                ...     ...          ...            ...   
32556   27            Private  257302   Assoc-acdm             12   
32557   40            Private  154374      HS-grad              9   
32558   58            Private  151910      HS-grad              9   
32559   22            Private  201490      HS-grad              9   
32560   52       Self-emp-inc  287927      HS-grad              9   

            marital-status          occupation    relationship    race  \
0            Never-married        Adm-clerical   Not-in-family   White   
1       Married-civ-spouse     Exec-managerial         Husband   White   
2                 Divorced   Handlers-cleaners   Not-in-family   White   
3       Married-civ-spouse   Handlers-cleaners         Husband   Black   
4       Married-civ-spouse      Prof-specialty            Wife   Black   
...                    ...                 ...             ...     ...   
32556   Married-civ-spouse        Tech-support            Wife   White   
32557   Married-civ-spouse   Machine-op-inspct         Husband   White   
32558              Widowed        Adm-clerical       Unmarried   White   
32559        Never-married        Adm-clerical       Own-child   White   
32560   Married-civ-spouse     Exec-managerial            Wife   White   

           sex  capital-gain  capital-loss  hours-per-week  native-country  \
0         Male          2174             0              40   United-States   
1         Male             0             0              13   United-States   
2         Male             0             0              40   United-States   
3         Male             0             0              40   United-States   
4       Female             0             0              40            Cuba   
...        ...           ...           ...             ...             ...   
32556   Female             0             0              38   United-States   
32557     Male             0             0              40   United-States   
32558   Female             0             0              40   United-States   
32559     Male             0             0              20   United-States   
32560   Female         15024             0              40   United-States   

      Listing of attributes  
0                     <=50K  
1                     <=50K  
2                     <=50K  
3                     <=50K  
4                     <=50K  
...                     ...  
32556                 <=50K  
32557                  >50K  
32558                 <=50K  
32559                 <=50K  
32560                  >50K  

[32561 rows x 15 columns]

In [5]:
adult_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   age                    32561 non-null  int64 
 1   workclass              32561 non-null  object
 2   fnlwgt                 32561 non-null  int64 
 3   education              32561 non-null  object
 4   eduaction-num          32561 non-null  int64 
 5   marital-status         32561 non-null  object
 6   occupation             32561 non-null  object
 7   relationship           32561 non-null  object
 8   race                   32561 non-null  object
 9   sex                    32561 non-null  object
 10  capital-gain           32561 non-null  int64 
 11  capital-loss           32561 non-null  int64 
 12  hours-per-week         32561 non-null  int64 
 13  native-country         32561 non-null  object
 14  Listing of attributes  32561 non-null  object
dtypes: int64(6), object

In [6]:
adult_data.isnull().sum()

age                      0
workclass                0
fnlwgt                   0
education                0
eduaction-num            0
marital-status           0
occupation               0
relationship             0
race                     0
sex                      0
capital-gain             0
capital-loss             0
hours-per-week           0
native-country           0
Listing of attributes    0
dtype: int64

In [7]:
print('age 값 분포',adult_data['age'].value_counts())

age 값 분포 36    898
31    888
34    886
23    877
35    876
     ... 
83      6
88      3
85      3
86      1
87      1
Name: age, Length: 73, dtype: int64


In [8]:
print('\n workclass 값 분포',adult_data['workclass'].value_counts())


 workclass 값 분포  Private             22696
 Self-emp-not-inc     2541
 Local-gov            2093
 ?                    1836
 State-gov            1298
 Self-emp-inc         1116
 Federal-gov           960
 Without-pay            14
 Never-worked            7
Name: workclass, dtype: int64


### workclass 변수중 ' ?' 제거

In [9]:
index = adult_data[adult_data['workclass']==' ?'].index

In [10]:
adult_data.drop(index, inplace = True)

In [11]:
print('\n fnlwgt 값 분포',adult_data['fnlwgt'].value_counts())


 fnlwgt 값 분포 203488    13
164190    13
123011    12
148995    12
121124    12
          ..
33669      1
177457     1
312767     1
43354      1
201490     1
Name: fnlwgt, Length: 20498, dtype: int64


In [12]:
print('\n education 값 분포',adult_data['education'].value_counts())


 education 값 분포  HS-grad         9969
 Some-college    6777
 Bachelors       5182
 Masters         1675
 Assoc-voc       1321
 11th            1057
 Assoc-acdm      1020
 10th             833
 7th-8th          574
 Prof-school      558
 9th              463
 Doctorate        398
 12th             393
 5th-6th          303
 1st-4th          156
 Preschool         46
Name: education, dtype: int64


In [13]:
print('\n eduaction-num 값 분포',adult_data['eduaction-num'].value_counts())


 eduaction-num 값 분포 9     9969
10    6777
13    5182
14    1675
11    1321
7     1057
12    1020
6      833
4      574
15     558
5      463
16     398
8      393
3      303
2      156
1       46
Name: eduaction-num, dtype: int64


In [14]:
print('\n marital-status 값 분포',adult_data['marital-status'].value_counts())


 marital-status 값 분포  Married-civ-spouse       14340
 Never-married             9917
 Divorced                  4259
 Separated                  959
 Widowed                    840
 Married-spouse-absent      389
 Married-AF-spouse           21
Name: marital-status, dtype: int64


In [15]:
print('\n occupation 값 분포',adult_data['occupation'].value_counts())


 occupation 값 분포  Prof-specialty       4140
 Craft-repair         4099
 Exec-managerial      4066
 Adm-clerical         3770
 Sales                3650
 Other-service        3295
 Machine-op-inspct    2002
 Transport-moving     1597
 Handlers-cleaners    1370
 Farming-fishing       994
 Tech-support          928
 Protective-serv       649
 Priv-house-serv       149
 Armed-Forces            9
 ?                       7
Name: occupation, dtype: int64


### occupation 변수중 ' ?' 처리

In [16]:
index = adult_data[adult_data['occupation']==' ?'].index

In [17]:
adult_data.drop(index, inplace = True)

In [18]:
print('\n occupation 값 분포',adult_data['occupation'].value_counts())


 occupation 값 분포  Prof-specialty       4140
 Craft-repair         4099
 Exec-managerial      4066
 Adm-clerical         3770
 Sales                3650
 Other-service        3295
 Machine-op-inspct    2002
 Transport-moving     1597
 Handlers-cleaners    1370
 Farming-fishing       994
 Tech-support          928
 Protective-serv       649
 Priv-house-serv       149
 Armed-Forces            9
Name: occupation, dtype: int64


In [19]:
print('\n relationship 값 분포',adult_data['relationship'].value_counts())


 relationship 값 분포  Husband           12704
 Not-in-family      7865
 Own-child          4525
 Unmarried          3271
 Wife               1435
 Other-relative      918
Name: relationship, dtype: int64


In [20]:
print('\n race 값 분포',adult_data['race'].value_counts())


 race 값 분포  White                 26301
 Black                  2909
 Asian-Pac-Islander      974
 Amer-Indian-Eskimo      286
 Other                   248
Name: race, dtype: int64


In [21]:
print('\n sex 값 분포',adult_data['sex'].value_counts())


 sex 값 분포  Male      20788
 Female     9930
Name: sex, dtype: int64


In [22]:
print('\n capital-gain 값 분포',adult_data['capital-gain'].value_counts())


 capital-gain 값 분포 0        28129
15024      343
7688       278
7298       244
99999      155
         ...  
6097         1
2538         1
401          1
1455         1
1086         1
Name: capital-gain, Length: 118, dtype: int64


In [23]:
print('\n capital-loss 값 분포',adult_data['capital-loss'].value_counts())


 capital-loss 값 분포 0       29257
1902      199
1977      167
1887      157
1848       50
        ...  
2457        1
4356        1
1539        1
1844        1
1411        1
Name: capital-loss, Length: 90, dtype: int64


In [24]:
print('\n hours-per-week 값 분포',adult_data['hours-per-week'].value_counts())


 hours-per-week 값 분포 40    14525
50     2763
45     1791
60     1441
35     1203
      ...  
82        1
94        1
92        1
87        1
74        1
Name: hours-per-week, Length: 94, dtype: int64


In [25]:
print('\n capital-loss 값 분포',adult_data['capital-loss'].value_counts())


 capital-loss 값 분포 0       29257
1902      199
1977      167
1887      157
1848       50
        ...  
2457        1
4356        1
1539        1
1844        1
1411        1
Name: capital-loss, Length: 90, dtype: int64


In [26]:
print('\n native-country 값 분포',adult_data['native-country'].value_counts())


 native-country 값 분포  United-States                 27504
 Mexico                          610
 ?                               556
 Philippines                     188
 Germany                         128
 Puerto-Rico                     109
 Canada                          107
 India                           100
 El-Salvador                     100
 Cuba                             92
 England                          86
 Jamaica                          80
 South                            71
 China                            68
 Italy                            68
 Dominican-Republic               67
 Vietnam                          64
 Guatemala                        63
 Japan                            59
 Poland                           56
 Columbia                         56
 Iran                             42
 Taiwan                           42
 Haiti                            42
 Portugal                         34
 Nicaragua                        33
 Peru           

In [27]:
index = adult_data[adult_data['native-country']==' ?'].index

In [28]:
adult_data.drop(index, inplace=True)

In [29]:
adult_data

age          workclass  fnlwgt    education  eduaction-num  \
0       39          State-gov   77516    Bachelors             13   
1       50   Self-emp-not-inc   83311    Bachelors             13   
2       38            Private  215646      HS-grad              9   
3       53            Private  234721         11th              7   
4       28            Private  338409    Bachelors             13   
...    ...                ...     ...          ...            ...   
32556   27            Private  257302   Assoc-acdm             12   
32557   40            Private  154374      HS-grad              9   
32558   58            Private  151910      HS-grad              9   
32559   22            Private  201490      HS-grad              9   
32560   52       Self-emp-inc  287927      HS-grad              9   

            marital-status          occupation    relationship    race  \
0            Never-married        Adm-clerical   Not-in-family   White   
1       Married-civ-spouse     Exec-managerial         Husband   White   
2                 Divorced   Handlers-cleaners   Not-in-family   White   
3       Married-civ-spouse   Handlers-cleaners         Husband   Black   
4       Married-civ-spouse      Prof-specialty            Wife   Black   
...                    ...                 ...             ...     ...   
32556   Married-civ-spouse        Tech-support            Wife   White   
32557   Married-civ-spouse   Machine-op-inspct         Husband   White   
32558              Widowed        Adm-clerical       Unmarried   White   
32559        Never-married        Adm-clerical       Own-child   White   
32560   Married-civ-spouse     Exec-managerial            Wife   White   

           sex  capital-gain  capital-loss  hours-per-week  native-country  \
0         Male          2174             0              40   United-States   
1         Male             0             0              13   United-States   
2         Male             0             0              40   United-States   
3         Male             0             0              40   United-States   
4       Female             0             0              40            Cuba   
...        ...           ...           ...             ...             ...   
32556   Female             0             0              38   United-States   
32557     Male             0             0              40   United-States   
32558   Female             0             0              40   United-States   
32559     Male             0             0              20   United-States   
32560   Female         15024             0              40   United-States   

      Listing of attributes  
0                     <=50K  
1                     <=50K  
2                     <=50K  
3                     <=50K  
4                     <=50K  
...                     ...  
32556                 <=50K  
32557                  >50K  
32558                 <=50K  
32559                 <=50K  
32560                  >50K  

[30162 rows x 15 columns]

In [30]:
adult_data.isnull().sum()

age                      0
workclass                0
fnlwgt                   0
education                0
eduaction-num            0
marital-status           0
occupation               0
relationship             0
race                     0
sex                      0
capital-gain             0
capital-loss             0
hours-per-week           0
native-country           0
Listing of attributes    0
dtype: int64

## adult_data_test

In [31]:
adult_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16281 entries, 0 to 16280
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   age                    16281 non-null  int64 
 1   workclass              16281 non-null  object
 2   fnlwgt                 16281 non-null  int64 
 3   education              16281 non-null  object
 4   eduaction-num          16281 non-null  int64 
 5   marital-status         16281 non-null  object
 6   occupation             16281 non-null  object
 7   relationship           16281 non-null  object
 8   race                   16281 non-null  object
 9   sex                    16281 non-null  object
 10  capital-gain           16281 non-null  int64 
 11  capital-loss           16281 non-null  int64 
 12  hours-per-week         16281 non-null  int64 
 13  native-country         16281 non-null  object
 14  Listing of attributes  16281 non-null  object
dtypes: int64(6), object

In [32]:
adult_test.isnull().sum()

age                      0
workclass                0
fnlwgt                   0
education                0
eduaction-num            0
marital-status           0
occupation               0
relationship             0
race                     0
sex                      0
capital-gain             0
capital-loss             0
hours-per-week           0
native-country           0
Listing of attributes    0
dtype: int64

In [33]:
print('age 값 분포',adult_test['age'].value_counts())

age 값 분포 35    461
33    460
23    452
36    450
31    437
     ... 
84      3
88      3
85      2
89      2
87      2
Name: age, Length: 73, dtype: int64


In [34]:
print('\n workclass 값 분포',adult_test['workclass'].value_counts())


 workclass 값 분포  Private             11210
 Self-emp-not-inc     1321
 Local-gov            1043
 ?                     963
 State-gov             683
 Self-emp-inc          579
 Federal-gov           472
 Without-pay             7
 Never-worked            3
Name: workclass, dtype: int64


In [35]:
index = adult_test[adult_test['workclass']==' ?'].index

In [36]:
adult_test.drop(index, inplace = True)

In [37]:
print('\n fnlwgt 값 분포',adult_test['fnlwgt'].value_counts())


 fnlwgt 값 분포 136986    9
203488    8
127651    8
120277    8
125892    8
         ..
326005    1
64102     1
385793    1
390537    1
83891     1
Name: fnlwgt, Length: 12081, dtype: int64


In [38]:
print('\n education 값 분포',adult_test['education'].value_counts())


 education 값 분포  HS-grad         5005
 Some-college    3261
 Bachelors       2590
 Masters          915
 Assoc-voc        657
 11th             577
 Assoc-acdm       509
 10th             408
 7th-8th          271
 Prof-school      252
 9th              224
 12th             206
 Doctorate        178
 5th-6th          165
 1st-4th           73
 Preschool         27
Name: education, dtype: int64


In [39]:
print('\n eduaction-num 값 분포',adult_test['eduaction-num'].value_counts())


 eduaction-num 값 분포 9     5005
10    3261
13    2590
14     915
11     657
7      577
12     509
6      408
4      271
15     252
5      224
8      206
16     178
3      165
2       73
1       27
Name: eduaction-num, dtype: int64


In [40]:
print('\n marital-status 값 분포',adult_test['marital-status'].value_counts())


 marital-status 값 분포  Married-civ-spouse       7112
 Never-married            4965
 Divorced                 2105
 Separated                 474
 Widowed                   456
 Married-spouse-absent     195
 Married-AF-spouse          11
Name: marital-status, dtype: int64


In [41]:
print('\n occupation 값 분포',adult_test['occupation'].value_counts())


 occupation 값 분포  Prof-specialty       2032
 Exec-managerial      2020
 Craft-repair         2013
 Sales                1854
 Adm-clerical         1841
 Other-service        1628
 Machine-op-inspct    1020
 Transport-moving      758
 Handlers-cleaners     702
 Tech-support          518
 Farming-fishing       496
 Protective-serv       334
 Priv-house-serv        93
 Armed-Forces            6
 ?                       3
Name: occupation, dtype: int64


In [42]:
index = adult_test[adult_test['occupation']==' ?'].index

In [43]:
adult_test.drop(index, inplace = True)

In [44]:
print('\n occupation 값 분포',adult_test['occupation'].value_counts())


 occupation 값 분포  Prof-specialty       2032
 Exec-managerial      2020
 Craft-repair         2013
 Sales                1854
 Adm-clerical         1841
 Other-service        1628
 Machine-op-inspct    1020
 Transport-moving      758
 Handlers-cleaners     702
 Tech-support          518
 Farming-fishing       496
 Protective-serv       334
 Priv-house-serv        93
 Armed-Forces            6
Name: occupation, dtype: int64


In [45]:
print('\n relationship 값 분포',adult_test['relationship'].value_counts())


 relationship 값 분포  Husband           6301
 Not-in-family     4051
 Own-child         2181
 Unmarried         1596
 Wife               704
 Other-relative     482
Name: relationship, dtype: int64


In [46]:
print('\n race 값 분포',adult_test['race'].value_counts())


 race 값 분포  White                 13143
 Black                  1447
 Asian-Pac-Islander      449
 Amer-Indian-Eskimo      149
 Other                   127
Name: race, dtype: int64


In [47]:
print('\n sex 값 분포',adult_test['sex'].value_counts())


 sex 값 분포  Male      10326
 Female     4989
Name: sex, dtype: int64


In [48]:
print('\n capital-gain 값 분포',adult_test['capital-gain'].value_counts())


 capital-gain 값 분포 0        14043
15024      165
7688       124
7298       114
99999       84
         ...  
2329         1
3273         1
2346         1
34095        1
2036         1
Name: capital-gain, Length: 111, dtype: int64


In [49]:
print('capital-gain 값 분포',adult_test['capital-gain'].value_counts())

capital-gain 값 분포 0        14043
15024      165
7688       124
7298       114
99999       84
         ...  
2329         1
3273         1
2346         1
34095        1
2036         1
Name: capital-gain, Length: 111, dtype: int64


In [50]:
print('hours-per-week 값 분포',adult_test['hours-per-week'].value_counts())

hours-per-week 값 분포 40    7242
50    1402
45     861
60     688
35     596
      ... 
74       2
89       1
76       1
79       1
69       1
Name: hours-per-week, Length: 89, dtype: int64


In [51]:
print('\n capital-loss 값 분포',adult_test['capital-loss'].value_counts())


 capital-loss 값 분포 0       14595
1902      102
1977       84
1887       73
2415       23
        ...  
1870        1
2282        1
1735        1
1825        1
1651        1
Name: capital-loss, Length: 80, dtype: int64


In [52]:
print('\n native-country 값 분포',adult_test['native-country'].value_counts())


 native-country 값 분포  United-States                 13788
 Mexico                          293
 ?                               255
 Philippines                      95
 Puerto-Rico                      66
 Germany                          65
 Canada                           56
 El-Salvador                      47
 India                            47
 China                            45
 Cuba                             41
 England                          33
 Italy                            32
 South                            30
 Japan                            30
 Dominican-Republic               30
 Portugal                         28
 Haiti                            27
 Columbia                         26
 Poland                           25
 Guatemala                        23
 Jamaica                          23
 Greece                           20
 Vietnam                          19
 Ecuador                          16
 Peru                             15
 Nicaragua      

In [53]:
index = adult_test[adult_test['native-country']==' ?'].index

In [54]:
adult_test.drop(index, inplace=True)

In [55]:
adult_test

age      workclass  fnlwgt      education  eduaction-num  \
0       25        Private  226802           11th              7   
1       38        Private   89814        HS-grad              9   
2       28      Local-gov  336951     Assoc-acdm             12   
3       44        Private  160323   Some-college             10   
5       34        Private  198693           10th              6   
...    ...            ...     ...            ...            ...   
16275   33        Private  245211      Bachelors             13   
16276   39        Private  215419      Bachelors             13   
16278   38        Private  374983      Bachelors             13   
16279   44        Private   83891      Bachelors             13   
16280   35   Self-emp-inc  182148      Bachelors             13   

            marital-status          occupation    relationship  \
0            Never-married   Machine-op-inspct       Own-child   
1       Married-civ-spouse     Farming-fishing         Husband   
2       Married-civ-spouse     Protective-serv         Husband   
3       Married-civ-spouse   Machine-op-inspct         Husband   
5            Never-married       Other-service   Not-in-family   
...                    ...                 ...             ...   
16275        Never-married      Prof-specialty       Own-child   
16276             Divorced      Prof-specialty   Not-in-family   
16278   Married-civ-spouse      Prof-specialty         Husband   
16279             Divorced        Adm-clerical       Own-child   
16280   Married-civ-spouse     Exec-managerial         Husband   

                      race      sex  capital-gain  capital-loss  \
0                    Black     Male             0             0   
1                    White     Male             0             0   
2                    White     Male             0             0   
3                    Black     Male          7688             0   
5                    White     Male             0             0   
...                    ...      ...           ...           ...   
16275                White     Male             0             0   
16276                White   Female             0             0   
16278                White     Male             0             0   
16279   Asian-Pac-Islander     Male          5455             0   
16280                White     Male             0             0   

       hours-per-week  native-country Listing of attributes  
0                  40   United-States                <=50K.  
1                  50   United-States                <=50K.  
2                  40   United-States                 >50K.  
3                  40   United-States                 >50K.  
5                  30   United-States                <=50K.  
...               ...             ...                   ...  
16275              40   United-States                <=50K.  
16276              36   United-States                <=50K.  
16278              50   United-States                <=50K.  
16279              40   United-States                <=50K.  
16280              60   United-States                 >50K.  

[15060 rows x 15 columns]

## IV. 인코딩

In [56]:
adult_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30162 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   age                    30162 non-null  int64 
 1   workclass              30162 non-null  object
 2   fnlwgt                 30162 non-null  int64 
 3   education              30162 non-null  object
 4   eduaction-num          30162 non-null  int64 
 5   marital-status         30162 non-null  object
 6   occupation             30162 non-null  object
 7   relationship           30162 non-null  object
 8   race                   30162 non-null  object
 9   sex                    30162 non-null  object
 10  capital-gain           30162 non-null  int64 
 11  capital-loss           30162 non-null  int64 
 12  hours-per-week         30162 non-null  int64 
 13  native-country         30162 non-null  object
 14  Listing of attributes  30162 non-null  object
dtypes: int64(6), object

#### I. 레이블 인코딩(train)

In [57]:
Label_adult_data= adult_data.copy()

In [58]:
Label_adult_data

age          workclass  fnlwgt    education  eduaction-num  \
0       39          State-gov   77516    Bachelors             13   
1       50   Self-emp-not-inc   83311    Bachelors             13   
2       38            Private  215646      HS-grad              9   
3       53            Private  234721         11th              7   
4       28            Private  338409    Bachelors             13   
...    ...                ...     ...          ...            ...   
32556   27            Private  257302   Assoc-acdm             12   
32557   40            Private  154374      HS-grad              9   
32558   58            Private  151910      HS-grad              9   
32559   22            Private  201490      HS-grad              9   
32560   52       Self-emp-inc  287927      HS-grad              9   

            marital-status          occupation    relationship    race  \
0            Never-married        Adm-clerical   Not-in-family   White   
1       Married-civ-spouse     Exec-managerial         Husband   White   
2                 Divorced   Handlers-cleaners   Not-in-family   White   
3       Married-civ-spouse   Handlers-cleaners         Husband   Black   
4       Married-civ-spouse      Prof-specialty            Wife   Black   
...                    ...                 ...             ...     ...   
32556   Married-civ-spouse        Tech-support            Wife   White   
32557   Married-civ-spouse   Machine-op-inspct         Husband   White   
32558              Widowed        Adm-clerical       Unmarried   White   
32559        Never-married        Adm-clerical       Own-child   White   
32560   Married-civ-spouse     Exec-managerial            Wife   White   

           sex  capital-gain  capital-loss  hours-per-week  native-country  \
0         Male          2174             0              40   United-States   
1         Male             0             0              13   United-States   
2         Male             0             0              40   United-States   
3         Male             0             0              40   United-States   
4       Female             0             0              40            Cuba   
...        ...           ...           ...             ...             ...   
32556   Female             0             0              38   United-States   
32557     Male             0             0              40   United-States   
32558   Female             0             0              40   United-States   
32559     Male             0             0              20   United-States   
32560   Female         15024             0              40   United-States   

      Listing of attributes  
0                     <=50K  
1                     <=50K  
2                     <=50K  
3                     <=50K  
4                     <=50K  
...                     ...  
32556                 <=50K  
32557                  >50K  
32558                 <=50K  
32559                 <=50K  
32560                  >50K  

[30162 rows x 15 columns]

In [59]:
from sklearn.preprocessing import LabelEncoder

features = Label_adult_data[['workclass','education','marital-status','occupation','relationship','race','sex','native-country','Listing of attributes']]
for feature in features:
    LE = LabelEncoder()
    LE = LE.fit(Label_adult_data[feature])
    Label_adult_data[feature] = LE.transform(Label_adult_data[feature])
    

In [60]:
Label_adult_data

age  workclass  fnlwgt  education  eduaction-num  marital-status  \
0       39          5   77516          9             13               4   
1       50          4   83311          9             13               2   
2       38          2  215646         11              9               0   
3       53          2  234721          1              7               2   
4       28          2  338409          9             13               2   
...    ...        ...     ...        ...            ...             ...   
32556   27          2  257302          7             12               2   
32557   40          2  154374         11              9               2   
32558   58          2  151910         11              9               6   
32559   22          2  201490         11              9               4   
32560   52          3  287927         11              9               2   

       occupation  relationship  race  sex  capital-gain  capital-loss  \
0               0             1     4    1          2174             0   
1               3             0     4    1             0             0   
2               5             1     4    1             0             0   
3               5             0     2    1             0             0   
4               9             5     2    0             0             0   
...           ...           ...   ...  ...           ...           ...   
32556          12             5     4    0             0             0   
32557           6             0     4    1             0             0   
32558           0             4     4    0             0             0   
32559           0             3     4    1             0             0   
32560           3             5     4    0         15024             0   

       hours-per-week  native-country  Listing of attributes  
0                  40              38                      0  
1                  13              38                      0  
2                  40              38                      0  
3                  40              38                      0  
4                  40               4                      0  
...               ...             ...                    ...  
32556              38              38                      0  
32557              40              38                      1  
32558              40              38                      0  
32559              20              38                      0  
32560              40              38                      1  

[30162 rows x 15 columns]

#### I. 레이블 인코딩(test)

In [61]:
Label_adult_test= adult_test.copy()

In [62]:
Label_adult_test

age      workclass  fnlwgt      education  eduaction-num  \
0       25        Private  226802           11th              7   
1       38        Private   89814        HS-grad              9   
2       28      Local-gov  336951     Assoc-acdm             12   
3       44        Private  160323   Some-college             10   
5       34        Private  198693           10th              6   
...    ...            ...     ...            ...            ...   
16275   33        Private  245211      Bachelors             13   
16276   39        Private  215419      Bachelors             13   
16278   38        Private  374983      Bachelors             13   
16279   44        Private   83891      Bachelors             13   
16280   35   Self-emp-inc  182148      Bachelors             13   

            marital-status          occupation    relationship  \
0            Never-married   Machine-op-inspct       Own-child   
1       Married-civ-spouse     Farming-fishing         Husband   
2       Married-civ-spouse     Protective-serv         Husband   
3       Married-civ-spouse   Machine-op-inspct         Husband   
5            Never-married       Other-service   Not-in-family   
...                    ...                 ...             ...   
16275        Never-married      Prof-specialty       Own-child   
16276             Divorced      Prof-specialty   Not-in-family   
16278   Married-civ-spouse      Prof-specialty         Husband   
16279             Divorced        Adm-clerical       Own-child   
16280   Married-civ-spouse     Exec-managerial         Husband   

                      race      sex  capital-gain  capital-loss  \
0                    Black     Male             0             0   
1                    White     Male             0             0   
2                    White     Male             0             0   
3                    Black     Male          7688             0   
5                    White     Male             0             0   
...                    ...      ...           ...           ...   
16275                White     Male             0             0   
16276                White   Female             0             0   
16278                White     Male             0             0   
16279   Asian-Pac-Islander     Male          5455             0   
16280                White     Male             0             0   

       hours-per-week  native-country Listing of attributes  
0                  40   United-States                <=50K.  
1                  50   United-States                <=50K.  
2                  40   United-States                 >50K.  
3                  40   United-States                 >50K.  
5                  30   United-States                <=50K.  
...               ...             ...                   ...  
16275              40   United-States                <=50K.  
16276              36   United-States                <=50K.  
16278              50   United-States                <=50K.  
16279              40   United-States                <=50K.  
16280              60   United-States                 >50K.  

[15060 rows x 15 columns]

In [63]:
from sklearn.preprocessing import LabelEncoder

features = Label_adult_data[['workclass','education','marital-status','occupation','relationship','race','sex','native-country','Listing of attributes']]
for feature in features:
    LE = LabelEncoder()
    LE = LE.fit(Label_adult_test[feature])
    Label_adult_test[feature] = LE.transform(Label_adult_test[feature])
    

In [64]:
Label_adult_test

age  workclass  fnlwgt  education  eduaction-num  marital-status  \
0       25          2  226802          1              7               4   
1       38          2   89814         11              9               2   
2       28          1  336951          7             12               2   
3       44          2  160323         15             10               2   
5       34          2  198693          0              6               4   
...    ...        ...     ...        ...            ...             ...   
16275   33          2  245211          9             13               4   
16276   39          2  215419          9             13               0   
16278   38          2  374983          9             13               2   
16279   44          2   83891          9             13               0   
16280   35          3  182148          9             13               2   

       occupation  relationship  race  sex  capital-gain  capital-loss  \
0               6             3     2    1             0             0   
1               4             0     4    1             0             0   
2              10             0     4    1             0             0   
3               6             0     2    1          7688             0   
5               7             1     4    1             0             0   
...           ...           ...   ...  ...           ...           ...   
16275           9             3     4    1             0             0   
16276           9             1     4    0             0             0   
16278           9             0     4    1             0             0   
16279           0             3     1    1          5455             0   
16280           3             0     4    1             0             0   

       hours-per-week  native-country  Listing of attributes  
0                  40              37                      0  
1                  50              37                      0  
2                  40              37                      1  
3                  40              37                      1  
5                  30              37                      0  
...               ...             ...                    ...  
16275              40              37                      0  
16276              36              37                      0  
16278              50              37                      0  
16279              40              37                      0  
16280              60              37                      1  

[15060 rows x 15 columns]

## V. 분류기(Decision Tree, Random Forest, Logistic Regression, SVM, GBM) 및 평가 지표

In [65]:
y_train = Label_adult_data['Listing of attributes']
X_train = Label_adult_data.drop('Listing of attributes',axis =1)
y_test = Label_adult_test['Listing of attributes']
X_test = Label_adult_test.drop('Listing of attributes',axis =1)

In [66]:
from sklearn.metrics import accuracy_score
# 판독기
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import GradientBoostingClassifier


# 결정트리, Random Forest, 로지스틱 회귀를 윟란 사이킷런 Classifier
dt_clf = DecisionTreeClassifier(random_state=11)
rf_clf = RandomForestClassifier(random_state=11)
lr_clf = LogisticRegression()
gb_clf = GradientBoostingClassifier(random_state=0)

# DecisionTreeClassifier 학습/예측/평가
dt_clf.fit(X_train, y_train)
dt_pred = dt_clf.predict(X_test)
print('DescisionTreeClassifier 정확도: {0: .4f}'.format(accuracy_score(y_test, dt_pred)))

# RandomForestClassifier 학습/예측/평가
rf_clf.fit(X_train, y_train)
rf_pred = rf_clf.predict(X_test)
print('RandomForestClassifier 정확도: {0: .4f}'.format(accuracy_score(y_test, rf_pred)))

# LogisticRegressionClassifier 학습/예측/평가
lr_clf.fit(X_train, y_train)
lr_pred = lr_clf.predict(X_test)
print('RogisticRegressionClassifier 정확도: {0: .4f}'.format(accuracy_score(y_test, lr_pred)))

# GBMClassifeir
gb_clf.fit(X_train, y_train)
gb_pred = gb_clf.predict(X_test)
print('GBMClassifier 정확도: {0: .4f}'.format(accuracy_score(y_test, gb_pred)))

DescisionTreeClassifier 정확도:  0.7991
RandomForestClassifier 정확도:  0.8505
RogisticRegressionClassifier 정확도:  0.7836
GBMClassifier 정확도:  0.8629


In [67]:
from sklearn.metrics import accuracy_score, precision_score , recall_score , confusion_matrix, f1_score 

def get_clf_eval(y_test , pred):
    confusion = confusion_matrix( y_test, pred)
    accuracy = accuracy_score(y_test , pred)
    precision = precision_score(y_test , pred)
    recall = recall_score(y_test , pred)
    f1 = f1_score(y_test,pred)
    print('오차 행렬')
    print(confusion)
    print('정확도: {0:.4f}, 정밀도: {1:.4f}, 재현율: {2:.4f}, F1:{3:.4f}'.format(accuracy, precision, recall, f1))


In [68]:
get_clf_eval(y_test,dt_pred)

오차 행렬
[[9775 1585]
 [1440 2260]]
정확도: 0.7991, 정밀도: 0.5878, 재현율: 0.6108, F1:0.5991


In [69]:
get_clf_eval(y_test,rf_pred)

오차 행렬
[[10566   794]
 [ 1458  2242]]
정확도: 0.8505, 정밀도: 0.7385, 재현율: 0.6059, F1:0.6657


In [70]:
get_clf_eval(y_test,lr_pred)

오차 행렬
[[10696   664]
 [ 2595  1105]]
정확도: 0.7836, 정밀도: 0.6246, 재현율: 0.2986, F1:0.4041


In [71]:
get_clf_eval(y_test,gb_pred)

오차 행렬
[[10787   573]
 [ 1491  2209]]
정확도: 0.8629, 정밀도: 0.7940, 재현율: 0.5970, F1:0.6816


## 1st. GBM 성능
정확도 : 0.8629  
정밀도 : 0.7940  
재현율 : 0.5970  
F1-Score : 0.6816

## 2st. Random Forest
정확도 : 0.8505  
정밀도 : 0.7385  
재현율 : 0.6059  
F1-Score : 0.6657